v2/internal/crdmanagement/crds.go

```diff
@@ -0,0 +1,340 @@
+	for _, goalCRD := range goal {
+
+		// Note that if the CRD is not found, we will get back a default initialized CRD.
+		// We run the comparators on that as they may match, especially if the comparator is something like
+		// "specs are not equal"
+		existingCRD := existingCRDs[goalCRD.Name]
+
+		// Deepcopy to ensure that modifications below don't persist
+		existingCRD = *existingCRD.DeepCopy()
+		goalCRD = *goalCRD.DeepCopy()
+
+		equal := true
+		for _, c := range comparators {
+			if c(existingCRD, goalCRD) == false {
+				equal = false
+				break
+			}
+		}
+
+		if equal {
+			matching[goalCRD.Name] = goalCRD
+		}
+	}
+
+	return matching
+}
+
+// FindNonMatchingCRDs finds the CRDs in "goal" that are not in "existing" OR are in "existing" but mismatch with the "goal"
+// based on the comparator functions.
+func (m *Manager) FindNonMatchingCRDs(
+	existing []apiextensions.CustomResourceDefinition,
+	goal []apiextensions.CustomResourceDefinition,
+	comparators ...func(a apiextensions.CustomResourceDefinition, b apiextensions.CustomResourceDefinition) bool,
+) map[string]apiextensions.CustomResourceDefinition {
+
+	// Just invert the comparators and call FindMatchingCRDs
+	invertedComparators := make([]func(a apiextensions.CustomResourceDefinition, b apiextensions.CustomResourceDefinition) bool, 0, len(comparators))
+	for _, c := range comparators {
+		c := c
+		invertedComparators = append(
+			invertedComparators,
+			func(a apiextensions.CustomResourceDefinition, b apiextensions.CustomResourceDefinition) bool {
+				return !c(a, b)
+			})
+	}
+
+	return m.FindMatchingCRDs(existing, goal, invertedComparators...)
+}
+
+func (m *Manager) ApplyCRDs(ctx context.Context, path string, podNamespace string, patterns []string) error {
+	goalCRDs, err := m.LoadOperatorCRDs(path, podNamespace)
+	if err != nil {
+		return err
+	}
+	m.logger.V(Info).Info("Goal CRDs", "count", len(goalCRDs))
+
+	existingCRDs, err := m.ListOperatorCRDs(ctx)
+	if err != nil {
+		return err
+	}
+	m.logger.V(Info).Info("Existing CRDs", "count", len(existingCRDs))
+
+	goalCRDsWithDifferentVersion := m.FindNonMatchingCRDs(existingCRDs, goalCRDs, VersionEqual)
+	goalCRDsWithDifferentSpec := m.FindNonMatchingCRDs(existingCRDs, goalCRDs, SpecEqual)
+
+	// The same CRD may be in both sets, but we don't want to apply twice, so combine the sets
+	crdsToApply := make(map[string]apiextensions.CustomResourceDefinition)
+	for name, crd := range goalCRDsWithDifferentVersion {
+		crdsToApply[name] = crd
+	}
+	for name, crd := range goalCRDsWithDifferentSpec {
+		crdsToApply[name] = crd
+	}
+
+	if len(crdsToApply) > 0 {
+		for _, crd := range crdsToApply {
+			crd := crd
+
+			_, isDifferentVersion := goalCRDsWithDifferentVersion[crd.Name]
+			_, isDifferentSpec := goalCRDsWithDifferentSpec[crd.Name]
+			m.logger.V(Verbose).Info("Found CRD in need of update", "CRD", crd.Name, "SpecDifferent", isDifferentSpec, "VersionDifferent", isDifferentVersion)
```

Comment:
For example, based on the above comment

```suggestion
			m.logger.V(Verbose).Info("Found CRD %d/%d in need of update", "CRD", 
			        index, count, crd.Name, "SpecDifferent", isDifferentSpec, "VersionDifferent", isDifferentVersion)
```

Reply:
Did it slightly differently than this because the logger wants us to put variables into the key/value parameters at the end. I did it like this:
```
			m.logger.V(Verbose).Info(
				"Found CRD in need of update",
				"Progress", fmt.Sprintf("%d/%d", i, len(crdsToApply)),
				"CRD", crd.Name,
				"SpecDifferent", isDifferentSpec,
				"VersionDifferent", isDifferentVersion)
```